### PR TITLE
Add Gitops deployment method

### PR DESCRIPTION
**Pull request by aperullo**
This PR adds support for a gitops approach to deploying grayskull. It will
- Install helm (v3) is only support version right now (PR coming soon to convert grayskull to helm v3)
- Deploys any secrets or crds that are needed for each role
- Templates out a collection of helmreleases/workloads on the kube-setup-delegate, following the Flux example repos folder structure 
- After the deploy key is manually added to the control repo, an admin can push the templated structure up to the control repo. The flux/helm operator will immediately start deploying the contents

Future work to be done:
- There is no gitops-remove or way to remove what gitskull installs
  - Flux **will** remove helm charts and workloads if they are removed from the repo. However it will **not** remove any crds that were installed as part of a chart.
- No auth role
- make gitskull push to a registry
- Deploy key must be added manually for now, it is printed during the run.

Issues: 
- Nodes can't pull from registry (Move following to an issue)
  - K8s reports errImagePull: no basic auth credentials
  - The config.json and ca.crt on gitops deployed (non-working) vs standard grayskull (working) is the same
  - Logging into a node and doing `docker pull` manually pulls the image without issue
  -  After manually pulling the image to the node, the pod will start up just fine

### CODE REVIEW

File: kubernetes/ansible/playbooks/gitskull.yml [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        - "*/LICENSE"
+        - "*/README.md"
+        - "*/tiller"
+      extra_opts:
+        - --strip-components=1
+
+  - name: Setup | Move helm binary into place
+    command: >
+      cp /tmp/linux-amd64/helm {{ bin_dir }}
+      creates={{ bin_dir }}/helm
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/gitskull.yml:78]: Not really part of this PR, but what does this do? Specifically, where does `/tmp/linux-amd64/helm` come from?

File: kubernetes/ansible/playbooks/gitskull.yml [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      vars:
+        task_modes: ["gitops-template"]
+    - include_role:
+        name: prometheus_role
+      vars:
+        task_modes: ["gitops-template"]
+    - include_role:
+        name: rook_ceph_role
+      vars:
+        task_modes: ["gitops-template"]
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/gitskull.yml:123]: Many of these tasks previously had an enabled option associated with them. Were they dropped intentionally or do you account for them somewhere else?

Comment (njbarber) [reply kubernetes/ansible/playbooks/gitskull.yml:123]: I guess a more specific question would be - is `gitskull.yml` designed to be run after `grayskull.yml` or to replace it?

File: kubernetes/ansible/playbooks/gitskull.yml [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  become: yes
+  tasks:
+    - include_role:
+        name: certs_role
+    - include_role:
+        name: gitops_role
+    - include_role:
+        name: cert_manager_role
+      vars: 
+        task_modes: ["gitops-install", "gitops-template"]
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/gitskull.yml:107]: Why does this task need a new `gitops-install` mode? Same question for the registry task below.

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+- name: Ingress | Run tasks
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/gitops_role/tasks/main.yml:1]: Name should be `Gitops`

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/remove.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    namespace: "{{ ingress_namespace }}"
+    kubectl: "{{ bin_dir }}/kubectl"
+    filename: "{{ role_dir }}/{{ item.item.file }}"
+    state: "absent"
+  with_items: "{{ ingress_role_manifests.results }}"
+
+- name: Ingress | Remove subfolder
+  file:
+    path: "{{ role_dir }}"
+    state: absent
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/gitops_role/tasks/remove.yml:50]: Name should be `Gitops` for all these as well.

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  file:
+    path: "{{ gitops_repo_dir }}/releases"
+    state: directory
+    mode: '0755'
+
+- name: Gitops | Create workloads folder
+  file:
+    path: "{{ gitops_repo_dir }}/workloads"
+    state: directory
+    mode: '0755'
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml:23]: Where did you find this recommended structure? I'm assuming this is what you're referring to in the PR description.

File: kubernetes/ansible/playbooks/roles/cert_manager_role/defaults/main.yml [hunk 7]
```diff
@@ -3,6 +3,13 @@ task_modes:
   - install
 
 cert_manager_namespace: grayskull-cert-manager
+gitops_namespace_dir: "{{ gitops_repo_dir }}/namespaces/{{ cert_manager_namespace }}"
+gitops_release_dir: "{{ gitops_repo_dir }}/releases/{{ cert_manager_namespace }}"
+gitops_workload_dir: "{{ gitops_repo_dir }}/workloads/{{ cert_manager_namespace }}"
+# Hacky solution to getting values file into HelmRelease
+# Read the values file, template, and indent it. Then assign it to a variable
+# When we template the helm release file the contents of this variable will be loaded
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/cert_manager_role/defaults/main.yml:11]: It looks like there are other ways to do this:

https://docs.fluxcd.io/projects/helm-operator/en/1.0.0-rc9/references/helmrelease-custom-resource.html#supplying-values-to-the-chart.

Is there a specific reason for this?

File: kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      - { file: dashboard-deployment.yml }
+      - { file: dashboard-certificate.yml }
+      - { file: dashboard-metrics-scraper.yml }
+
+- name: Dashboard | Template manifests
+  template:
+    src: "{{ item.file }}"
+    dest: "{{ gitops_workload_dir }}/{{ item.file }}"
+  with_items: "{{ dashboard_role_templates }}"
+  register: dashboard_role_manifests
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: Why is there a separation between namespaces and workloads? If you change the name of the namespace, for instance, you will still have workloads which reference the existing namespace. What is the anticipated behavior here?

File: kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    src: elk-namespace.yml
+    dest: "{{ gitops_namespace_dir }}/elk-namespace.yml"
+
+#---- Need to get the values into the release but don't want to duplicate them in case values file changes.
+
+- name: ELK | Create subfolder
+  file:
+    path: "{{ role_dir }}"
+    state: directory
+    mode: '0755'
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml:30]: Is this necessary? You don't have it for some of the other `gitops-template` tasks.

File: kubernetes/ansible/playbooks/roles/cert_manager_role/tasks/gitops-template.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    path: "{{ gitops_workload_dir }}"
+    state: directory
+    mode: '0755'
+
+- name: Cert-manager | Template namespace
+  template:
+    src: cert-manager-namespace.yml
+    dest: "{{ gitops_namespace_dir }}/cert-manager-namespace.yml"
+
+- name: Cert-manager | Tempate custom resource definitions
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/cert_manager_role/tasks/gitops-template.yml:24]: Template

File: kubernetes/ansible/playbooks/gitskull.yml (hunk 1 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/gitskull.yml:78]: I'm not sure... I checked the directory but its empty. This might be part of an old way of doing it.

File: kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      - { file: logstash-release.yml }
+
+- name: ELK | Template releases
+  template:
+    src: "gitops/{{ item.file }}"
+    dest: "{{ gitops_release_dir }}/{{ item.file }}"
+  with_items: "{{ elk_values_templates }}"
+  register: elk_values_manifests
+
+- name: ELK | Template release
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml:46]: This isn't a release.

File: kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+#---- Need to get the values into the release but don't want to duplicate them in case values file changes.
+
+- name: ELK | Create subfolder
+  file:
+    path: "{{ role_dir }}"
+    state: directory
+    mode: '0755'
+
+- name: ELK | Roles list
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/elk_role/tasks/gitops-template.yml:32]: Is roles the right word here?

File: kubernetes/ansible/playbooks/gitskull.yml (hunk 3 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/gitskull.yml:107]: For creating the secrets used in the issuers. Can't be part of the stuff uploaded to it.

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/main.yml (hunk 4 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/main.yml:1]: Good catch

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/remove.yml (hunk 5 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/remove.yml:50]: The remove playbook doesn't do anything yet. I probably should have just deleted it. I'm not sure how to go about making one that does work. If I just delete the flux operator than whatever it has deployed from git will stick around and leave the cluster is a half broken state.

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml (hunk 6 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml:23]: Its modified from the [flux-get-started](https://github.com/fluxcd/flux-get-started) repo. I added a folder per namespace because that made the ansible tasks more uniform, but otherwise it follows that structure

File: kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml (hunk 8 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: If you change the namespace name, retemplating with ansible should update all the references unless I missed a spot.

Comment (njbarber) [reply kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: Not sure I understand. Aren't you templating all these resources and then committing to the control repo?

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml (hunk 6 above)
Comment (njbarber) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml:23]: I'm having some trouble identifying what goes where. You have two different directories when rendered with defaults - `grayskull/gitops-role` and `grayskull/gitops-repo`, the latter representing the control repo. It looks like everything is copied into `grayskull/gitops-role`, which I get since it wouldn't make any sense to have what's controlling gitops in a control repo, but you create folders for namespaces, releases, and workloads under `grayskull/gitops-repo`. What exactly would go in these?

File: kubernetes/ansible/playbooks/gitskull.yml (hunk 2 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/gitskull.yml:123]: They are independent. You run one _or_ the other.

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml (hunk 6 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml:23]: The gitops role doesn't have anything that needs to go to the control repo since its the one reading it. All of its config is done by ansible alone, so I use our usual pattern of copying to the `role_dir`. Most of the other roles only need things deployed via gitops, so they put their products into `gitops-repo`, which is the directory to be uploaded to the control repo

File: kubernetes/ansible/playbooks/roles/cert_manager_role/defaults/main.yml (hunk 7 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/cert_manager_role/defaults/main.yml:11]: After some discussion, switching to one of the alternate ways of doing it won't solve the core problem of the hacky way of templating, just move it elsewhere. This is because I would still need to load a templated values into a template configmap, which would have to employ the same slightly hacky solution.

File: kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml (hunk 8 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: Yeah

File: kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml (hunk 6 above)
Comment (njbarber) [reply kubernetes/ansible/playbooks/roles/gitops_role/tasks/gitops-template.yml:23]: That's what I figured but wanted to confirm.

File: kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml [hunk 13]
```diff
@@ -6,7 +6,8 @@
 
 - name: Ingress | Download chart
   shell: "{{ bin_dir }}/helm fetch \
+    --repo {{ helm_ingress_repo_url }} \
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:9]: It looks like this is guaranteed to already be added:

https://github.com/connexta/grayskull/pull/111/files#diff-ac117f9344befa980a42c5062301b642R81

But most of the other download tasks have `--repo` as well. Just thought I'd point it out. What you have here is probably better in the event the repo changes.

File: kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml [hunk 14]
```diff
@@ -6,7 +6,8 @@
...
     --version {{ helm_ingress_chart_version }} \
     --destination {{ role_dir }} \
     --untar \
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:12]: Is this necessary? It looks like you created new download tasks without it:

https://github.com/connexta/grayskull/pull/111/files#diff-b7e851a84cfaf93f6ee4f4750b39c1bfR14-R19

Comment (njbarber) [reply kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:12]: We had discussed this previously. If I remember correctly, Helm doesn't care if the destination is a tarball. If so, we should adopt the approach you used above and remove `--untar` from all the other download tasks.

File: kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml (hunk 13 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:9]: I did this to make it uniform across all the roles. No reason to have `chart_type` instead of  `repo_url` in only 2 places, and I think I remember helm3 doing away with that convention anyway. Flux also prefers using the url so this change accommodates that.

File: kubernetes/ansible/playbooks/roles/ion_role/tasks/gitops-template.yml [hunk 15]
```diff
@@ -0,0 +1,28 @@
+- name: Ion | Create namespace folder
+  file:
+    path: "{{ gitops_namespace_dir }}"
+    state: directory
+    mode: '0755'
+
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/ion_role/tasks/gitops-template.yml:6]: I'm not that familiar with what had been done for KR staging, so you may already have a plan for this, but how should the ion repos tie into this i.e. should there be a helm release for each one and, if so, should there be a release folder here to accommodate that?

File: kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml (hunk 14 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:12]: > adopt the approach you used above
did I remove `--untar` somewhere in this pr? I must have missed that.

File: kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml (hunk 8 above)
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: It doesn't look like you're using `dashboard_role_manifests` anywhere.

Comment (njbarber) [reply kubernetes/ansible/playbooks/roles/dashboard_role/tasks/gitops-template.yml:43]: Same question for these:

https://github.com/connexta/grayskull/pull/111/files#diff-ffb5619c384da56eec6aa1ea744350f9R38
https://github.com/connexta/grayskull/pull/111/files#diff-31243b0aeb3cf22ea82191248ef36dc7R28
https://github.com/connexta/grayskull/pull/111/files#diff-bbc9235b4ea4c30e1323e54c45e99673R44
https://github.com/connexta/grayskull/pull/111/files#diff-1077cd53d47cdc5f6d53f8b01fb7e41aR52

File: kubernetes/ansible/playbooks/roles/registry_role/tasks/gitops-template.yml [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+      path: "{{ gitops_workload_dir }}"
+      state: directory
+      mode: '0755'
+
+  - name: Registry | Template namespace
+    template:
+      src: registry-namespace.yml
+      dest: "{{ gitops_namespace_dir }}/registry-namespace.yml"
+
+  - name: Registry | Template namespace
```
Comment (njbarber) [inline kubernetes/ansible/playbooks/roles/registry_role/tasks/gitops-template.yml:28]: Template certificate

File: kubernetes/ansible/playbooks/gitskull.yml (hunk 1 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/gitskull.yml:78]: It was. Removed it

File: kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml (hunk 14 above)
Comment (aperullo) [reply kubernetes/ansible/playbooks/roles/ingress_role/tasks/download.yml:12]: Changed in the next PR
